### ISSUE TITLE
Documentation online

### ISSUE THREAD

**Issue by zdenop** [DanBloomberg/leptonica#126]
Can you please publish recent documentation of leptonica?
Something like https://tpgit.github.io/UnOfficialLeptDocs/leptonica/ or https://tpgit.github.io/UnOfficialLeptDocs/leptonica/functions.html or doxygen generated (like tesseract-ocr.github.io)

**Comment by DanBloomberg** [comment]
We don't have a recent update to the UnOfficial documentation.

However, documentation continues to be updated at leptonica.org.

  -- Dan

**Comment by pullmoll** [comment]
I already thought about giving a try to change the existing documentation blocks for all the functions to the [doxygen](http://www.stack.nl/~dimitri/doxygen/) format. This would make it very much easier to keep some online documentation up to date.

**Comment by DanBloomberg** [comment]
Jürgen,

What is the simplest change that can reasonably be made to augment the
current doxygen formatting?

At present, nearly every function has a text header
   /*!
      ....
   */

Is there any reason to add anything?

With respect to the comments at the top of each file, these do two things:
  (1) list the functions in the file
  (2) [sometimes] have other general comments about usage

These are presently not 'doxygenated'.

Would you suggest using the following to include them as well:

   /*!  \file
      ....
   */

Should there be any special formatting for the listing of the functions?

Perhaps you can take a file, say, boxbasic.c (which has a listing of
functions and a paragraph on usage) and decorate it, as an example of what
you are thinking about?

  -- Dan

**Comment by pullmoll** [comment]
Hey Dan,

I don't know, if the input parameters and return values would be detected by doxygen as is. I rather think the entire comment block would be taken as is, but without specific distinction of the input and return values. I'm no expert with doxygen, though.

I'm using QtCreator as my preferred editor for C/C++ files and it offers a nice shortcut to complete a doxygen comment, if you just type the three characters `/**` and `<ENTER>` in the line above a function. It then inserts a template comment block where every parameter is inserted like `* @param name` and the return value as `* @return`, just as in my example above.

The existing comments are preciously formatted in the numbered sections, which I would try to keep for the output with the `<pre></pre>` HTML tags.

In addition, instead of the function's name (which is obviously the one following the comment) there should be a `* @brief ` description of what it does.

Adding `\file` or `@file` descriptions is certainly a good idea. I think there will also be a way to somehow group functions into sections, which is currently done by header lines inside the file comment.

I have to RTFM of doxygen to see how to do it right. I believe there is some doxygen specific notation for input vs. output parameters and optional params as well. The current `<output>` or `<optional>` marks would probably and wrongly be taken as HTML tags.

Also there should be a global configuration for doxygen (e.g. Doxyfile) to define the details of the generated output.

The most part of a conversion, however, would be to put the `@param` (or `\param`) in front of all parameters and `@return` before the return values of every function, plus the HTML tagging of the `Note:` sections as preformat. 

As an example I've done this for [adaptmap.c](http://sprunge.us/CUNE), the alpabetically first source file. I used QtCreators keyboard macros and some manual work to do it.

**Comment by pullmoll** [comment]
I hacked up a comment format converter which should do most of the tedious part:
https://github.com/pullmoll/leptonica/tree/doxygen

What's missing is the detection of the file block comments and insertion of `\file` in order to make doxygen actually list the files and their functions in its index.

**Comment by DanBloomberg** [comment]
Wow.  That was/is a big job!  I really appreciate your effort here.

I have only looked at a few files.  I like to use block headers such as

/*-------------------------------------------------------------------------*
 *                     Regression test parameter packer            *
 *-------------------------------------------------------------------------*/

but I see that in regutils.h, it was replaced by

 /*! Regression test parameter packer */

Perhaps we can do:

 /*-------------------------------------------------------------------------*/
 /*!                   Regression test parameter packer            */
 /*-------------------------------------------------------------------------*/

I hope you're not suggesting 2600 TODOs to write "brief" descriptions of each function.  That just can't happen -- life is too short    :-)  
Basically, if I believe the function and/or parameters need to be further explained, I add the Notes: section.  Otherwise, the one-liners following the input args suffice.

What do you suggest at this time?  That we pull the entire change as is (maybe with some modificatoins as mentioned above?)

**Comment by DanBloomberg** [comment]
Jurgen, how do I pull your changes into a local copy, rather than into the
master?   (I'm just a novice with git/github).

  -- Dan

On Sat, Apr 9, 2016 at 11:25 AM, Jürgen Buchmüller <notifications@github.com

**Comment by pullmoll** [comment]
Dan, you can perhaps create a branch on your side with `git checkout -b doxygen` and then `git pull  git@github.com:/pullmoll/leptonica -b doxygen`. If that fails, the `git pull --help` should tell how to do it right :)

I'd first have to create a PR from my branch to make it easier. I would not yet create it, because I think I can find some more bugs and make some more improvements to the converter.

I will reset the header files to the original and modify them in a way that the blocks are kept.

For the TODOs I wanted to try to collect the previous block comment before a function and write that to the lookup file. I don't know how many useful entries this might generate.

**Comment by pullmoll** [comment]
I rewrote the header files without removing the
```
/*******************************
 *      Block comments     *
 *******************************/
```
The tagging for doxygen with `*!` seems to work as [can be seen here](http://misc.voidlinux.de/html/pix_8h.html).

**Comment by pullmoll** [comment]
I'm still working on improving the results. The current patchset (before `conv2doxy` is run) can be [inspected here](https://github.com/DanBloomberg/leptonica/compare/master...pullmoll:doxygen?expand=1). Once I'm satisfied with the results, I'll actually open up the PR.

I'm trying hard not to change too much, but just what's required to allow a smooth transition by running `conv2doxy` at some point.

**Comment by DanBloomberg** [comment]
Looks good!  Agree -- less is better, unless it's too little  :-)

I am thinking that once this is set up, we could run your converter once
with each new version (1.73, ...).  Does that seem reasonable?

Is it too much work to add the
    ! \file
before the list of functions at the top of each .c file?

Actually, if you get this in shape there's no reason for me to integrate
and edit it before pushing to the master.

  -- Dan

On Mon, Apr 11, 2016 at 2:50 AM, Jürgen Buchmüller <notifications@github.com

**Comment by pullmoll** [comment]
I think that perhaps you did not look at the files *after* conv2doxy was run, as they are not online. In my pull request there are just format, alignment and missing/extra pieces fixes to the src files *before* conv2doxy is run.

So, NOOO! :) You would run the converter exactly __once__ at some point and afterwards check in the converted `src/*.c` files. In the future, if new functions are added, you would use the doxygen comment style yourself. That makes it much easier for contributors IMO, because doxygen is very well documented (who'da thought) and following its rules is simple.

Also everyone with a github account can then help fixing or improving the documentation through pull requests without the need to re-run some other tool but doxygen. Perhaps the Makefile.am can be modified to have a target `doc` which runs doxygen.

The `/*! \file ...` lines are inserted. That's part of the one-time conversion process.

I could create the PR now, because my recent changes were all just fixes to e.g. mismatches between parameter names in the comment vs. the function header, or some kind of copy+paste errors. They are all documented in the doxygen log output and can be fixed any time after the conversion.

**Comment by DanBloomberg** [comment]
You are right -- I haven't looked at the files after conv2doxy.  But I do
want to (see below).

Question: did you insert the /*! \file labels by hand on all the files?
That's not part of conv2doxy?

I haven't been able to test your doxygen branch.  I did 'git checkout -b
doxygen', but the 'git pull git@github.com:/pullmoll/leptonica' doesn't do
anything (it doesn't include your patch), and appending the '-b doxygen'
isn't legal.  I also tried a git pull on the URL of your leptonica
directory, but that gave the same result -- I didn't get the doxygen
branch.  And I couldn't figure out what to do from 'git pull --help'.

I just want to download your patch and play with it, run your conv2doxy on
a few files, etc.  This seems like a basic thing to do (e.g., before
merging with the master on a big change).  Any more suggestions on how I
can recreate your doxygen branch locally?

  -- Dan
